lnst/RecipeCommon/endpoints.py

```diff
@@ -85,3 +85,17 @@ def from_devices_with_ips(
             IPEndpoint(dev1, selected_ips[0]),
             IPEndpoint(dev2, selected_ips[1]),
         )
+
+
+def sequential_ip_endpoints(
+    *device_pairs: tuple[Device, Device]
+) -> list[list[EndpointPair[IPEndpointLike]]]:
+    """Helper function for use in generate_perf_endpoints method.
+
+    Generates a sequential endpoint list for each input device pair.
+    """
+    return [
+        [EndpointPair.from_devices_with_ips(dev1, dev2, ip_type=ip_type)]
```

Comment:
this inner list is here due to the internal semantics of the `generate_perf_endpoints` right? if so then i think it should be removed from here as this is a more "common/generic" module and instead the `generate_perf_endpoints` method should then do an additional transformation which adds the use case specific semantics.

Reply:
So, you're proposing changing each recipe's `generate_perf_endpoints` like this:
```py
# before
return sequential_ip_endpoints(
    (self.matched.host1.eth0, self.matched.host2.eth0),
)

# after
dev1, dev2 = self.matched.host1.eth0, self.matched.host2.eth0
return [
    [EndpointPair.from_devices_with_ips(dev1, dev2, ip_type=ip_type)]
    for ip_type in [Ip4Address, Ip6Address]
]
```
Can't we just move the function to a different module instead?

Reply:
well my "internal semantics of `generate_perf_endpoints`" i meant the method or how it is used - we can do your proposal or move the filter into the function that USES the result of the `generate_perf_endpoints` so imo this would be fine as well:

```python
return sequential_ip_endpoints((self.matched.host1.eth0, self.matched.host2.eth0))
```

and then the user of the method:
```python
useful_endpoints = apply_some_filters(self.generate_perf_endpoints)
```

Reply:
That's kind of how this is done now. The filtering is done in `BaseFlowMeasurementGenerator`'s `generate_flow_combinations()`

Reply:
Changed it so this method returns a simple list of generated endpoints. Since it's now strongly tied to `BaseEnrtRecipe`, we should maybe move it there, maybe even make it a method of `EnrtConfiguration`?

Reply:
ah i see you've made the same comment i have independently while reviewing this. i'll copy that comment here and close the other thread so that we can keep the context in just a single thread:



i don't really like that this lnst.RecipeCommon.endpoints function is importing an lnst.Recipes.ENRT.BaseEnrtRecipe class... the recipe common package should be a more "common" supporting library that is used in the Recipes package. this is in reverse, creating a sort of cycle... could actually easily lead to cyclical dependencies imo.

i think it would be safer:

    move this method into the enrt package - if it's a helper method that we strictly use in the enrt package, and it depends on using classes/types from the enrt package... it should be in the enrt package
    if it is a more common use case method, it should use more common use case arguments/types. and so the method signature should be more like ip_endpoints(*device_pairs: tuple[tuple[Device, list[IpAddress]], tuple[Device, list[IpAddress]]])

Reply:
I'll move it elsewhere. I don't really want to make it more universal, because each recipe would have to do this:
```py
# before
ip_endpoints(config, (dev1, dev2))

# after
ip_endpoints(((dev1, config.ips_for_device(dev1)), (dev2, config.ips_for_device(dev2))))
```

Reply:
Moved to `lnst.Recipes.ENRT.helpers`.

Reply:
looks ok